### ISSUE TITLE
What are the current problems with Badge?

### ISSUE THREAD

**Issue by hmstepanek** [hypothesis/product-backlog#953]
This spike is done when:

- [x] We've listed all the proposed problems with the badge
- [x] We've gone over the list of problems and decided which are problems we want to address in the short-term, and which we can <del>strike out</del> because we don't think they're actually important problems or don't want to address them in the short-term
- [ ] We've put the surviving list of problems into priority order

The output of this spike will be:

- [ ] A priority ordered list of problems with the badge endpoint that we actually want to solve in the short term

# The Problems

Some of the possible problems that have been proposed (some of these may not actually be important problems):

1. **User privacy concern:** Our Chrome extension sends every URL the user's browser visits to our servers. Regardless of whether we keep these URLs or not, that's a privacy concern to users.

   - Dan and Arti have said that this isn't actually a concern

   2019/02/12 seems like the team would still prefer to address this issue. Partially because it seems it should be easy to do something about.

2. <del>**h performance:** The badge is currently 72.5% of requests to h, and does not scale at the same rate as the rest of H since it's not linked to usage of our service. Reducing the badge request load to h ought to make h's current performance at least _somewhat_ better and make debugging h performance issues easier.</del>

   <del>- _Is_ the badge actually a big impact on h performance?</del>

   <del>- Badge searches currently account for about 14% of all searches in h (at least 1% of all badge requests result in a search). We are reaching our limit with our current elasticsearch resources at peak traffic and if we were able to decrease the searches from badge we would (probably?) be able to prolong scaling our Es cluster.</del>

   2019/02/12 The team decide to strike this out as an issue to be addressed in the short term because we _can't_ address it. Given that the Elasticsearch queries resulting from badge hits (cases when the badge number is > 0) are responsible for most of the performance issue, and it's not possible to eliminate those badge hit Elasticsearch queries without changing the UX of the badge, we will not be able to address this without changing UX. I think that if we really, really wanted to address this without making UX changes we _might_ be able to come up with some way. But we decided to strike this one out.

3. **Cost:** Badge costs us 60% of our total bill and isn't doesn't correlate with user's using our service.

   - On the other hand it doesn't cost us much in development time. The badge is a simple feature that we haven't to make many changes to or fix many bugs in, that is not a separate service we have to maintain, etc.

   2019/02/19 Decided to make simple Chrome extension changes only, limit cost reduction attempts to that.

4. <del>**Scalability:** The badge isn't scalable. One request per URL browsed to by anyone with our Chrome extension installed, even if they aren't logged in to us. If we one day have 10x as many users, 100x, a billion users, etc, that's not gonna work.</del>

   - <del>_Do_ we actually need to scale up the badge endpoint? Only Chrome extension users use the badge. Users of publishers with the client embedded, LMS users, and Via users don't make badge requests. The user growth we're currently pursuing is in publisher embeds and the LMS app, not in Chrome extension installs, and I don't think our number of Chrome extension users is currently growing dramatically, so this may not be an immediate problem for us.</del>

   Anything we implement should be designed to be _somewhat_ scalable but we don't want to solve the big badge problem right now. So if we have ~100,000 Chrome extension users right now, we should design something we think will work with 1M users. And similar scalability for num. URLs in DB.

5. <del>**UX:** Current effectiveness of badge features are not measurable (meaning we don't know whether badge is effective in getting user's to annotate, whether they are using the feature, etc) because it's an opt-out feature and has no measurable association with the user opening the sidebar to annotate. For example: badge runs on every url whether the user is logged in, using the sidebar or not, so how do we tell when the user opened the sidebar because of badge vs because they simply wanted to annotate?</del>

   2019/02/12 Team decided not to address this issue in the short term

6. <del>**Bugs:** The badge count doesn't always match what the user sees when they open the sidebar so effectively it appears buggy. We've had several customers report problems with badge counts being incorrect. </del>

   2019/02/12 Team decided to deal with these as separate issues

7. **h service purity:** A "badge endpoint" is Chrome extension-specific and doesn't belong in our universal annotation service.

   - The badge has different scalability properties than the rest of h. Scales with num. Chrome extension installs, not with Chrome extension or client usage.
   - Does it really belong in some sort of metadata / stats service?

   2019/02/19 Break out of h because of the scaling difference. But note that we don't think that it's a problem to have computing the badge number itself in h.

   It's ok to move only the badge misses, not the hits, out of h, because there are fewer hits and they probably scale more similarly to the rest of h's traffic.

8. **Sentry noise** and general monitoring noise

   2019/02/19  Given that we decided to address (7) we do want to address this as well since a solution to (7) will probably solve (8) too

**Comment by seanh** [comment]
Dev team has a meeting about this today. We discussed problems 1 (privacy), 2 (h performance), 5 (UX) and 6 (bugs). We struck out 2, 5 and 6, but we decided that 1 (privacy) **is** a problem that we want to address in the short-term.

We need to arrange another meeting to discuss problems 3, 4, 7 and 8 and to prioritise the surviving problems and then we should be able to close this spike.

**Comment by seanh** [comment]
Another dev team meeting: We finished discussing the 8 problems above and decided and 4 problems that we do want to address in the short-term. We created three further spike issues to investigate solutions to those 4 problems (2 of the problems are combined into 1 spike).

For each of the spikes, once the spike is completed and a solution has been outlined and agreed on, then we could create a further card to implement and deploy that solution.

1. [Outline a solution to the badge user privacy issue](https://github.com/hypothesis/product-backlog/issues/960)
2. [Outline simple ways to reduce the AWS costs of the badge](https://github.com/hypothesis/product-backlog/issues/961)
3. [Outline a way to remove "badge misses" from h](https://github.com/hypothesis/product-backlog/issues/962)
